### ISSUE TITLE
Deserialization Bug

### ISSUE THREAD

**Issue by ccaapton** [pyparsing/plusminus#22]
I'm trying to use this project in a spark distributed environment, but encountered a "RecursionError: maximum recursion depth exceeded". After some debugging, I found this is because the parser and parsed_rule can not be de-serialized.

Below is a sample code to reproduce the error:
```
import cloudpickle
from plusminus import ArithmeticParser

parser = ArithmeticParser()
parsed_res = parser.parse(' A + B')
parsed_pair = cloudpickle.dumps((parser, parsed_res))
parser, parsed_res = cloudpickle.loads(parsed_pair)
```

**Comment by ptmcg** [comment]
Thanks for pointing me toward cloudpickle. I have done similar work using pickle and dill in the past, and added features to pyparsing (on which plusminus is based) to improve support for pickling.

To verify that pyparsing itself works with cloudpickle, I took your example, but had it just parse a simple pyparsing parser (though I added a few tricky parts, a parse action lambda and named results). Here is that test code:

```python
import sys

import cloudpickle
import pyparsing as pp

# define a parser, including a term with a parse action, and results names
parser = pp.Word(pp.alphas)("word") + pp.Word(pp.nums).add_parse_action(lambda t: int(t[0]))("num")
parsed_res = parser.parse_string("abc 123")
print(parsed_res.dump())

# pickle the parser and the parsed results
parsed_pair = cloudpickle.dumps((parser, parsed_res))

# unpickle parser and results, and print unpickled results
print("\nUnpickled")
parser_unpickled, parsed_res_unpickled = cloudpickle.loads(parsed_pair)
print(parsed_res_unpickled.dump())

# Make sure unpickled parser works okay, even after removing pyparsing import
# from sys.modules
del pp
sys.modules.pop("pyparsing")
print(parser_unpickled.parse_string("def 456").dump())
```
giving:
```
['abc', 123]
- num: 123
- word: 'abc'

Unpickled
['abc', 123]
- num: 123
- word: 'abc'
['def', 456]
- num: 456
- word: 'def'
```
which is exactly what I expected.

Then I went a step further and created a small recursive parser that is very similar to what we see in plusminus, a parser for infix notation expressions.

```python
term = pp.Word(pp.nums).add_parse_action(lambda t: int(t[0])) | pp.Word(pp.alphas)
expr = pp.infix_notation(
    term,
    [
        ("+", 2, pp.opAssoc.LEFT),
    ]
)
parsed_expr = expr.parse_string("A + B + 1000")
print(parsed_expr.dump())

parsed_pair = cloudpickle.dumps((expr, parsed_expr))
expr_unpickled, parsed_expr_unpickled = cloudpickle.loads(parsed_pair)
print(parsed_expr_unpickled.dump())
print(expr_unpickled.parse_string("C + D + 2000").dump())
```

This also produces successful and expected results:
```
[['A', '+', 'B', '+', 1000]]
[0]:
  ['A', '+', 'B', '+', 1000]
[['A', '+', 'B', '+', 1000]]
[0]:
  ['A', '+', 'B', '+', 1000]
[['C', '+', 'D', '+', 2000]]
[0]:
  ['C', '+', 'D', '+', 2000]
```

Given these quick experiments, I'm not sure why plusminus's parser or parsed results fail with cloudpickle. This may require engaging with those working on cloudpickle to see if they have more insights, or if they have seen that recursion exception before.

**Comment by ccaapton** [comment]
I also notice that pyparsing itself will not cause this issue, but plusminus does. I modify the __getattr__ of ArithmeticParser a bit and add a print here as below. You can see that it is caused by "parser = self._parser" line.

```
import cloudpickle
from plusminus import ArithmeticParser

class TestClass(ArithmeticParser):
    def __getattr__(self, attr):
        print(attr)
        parser = self._parser
        if hasattr(parser, attr):
            return getattr(parser, attr)
        raise AttributeError(f"no such attribute {attr!r}")

parser = TestClass()
parsed_res = parser.parse(' A + B')
parsed_pair = cloudpickle.dumps((parser, parsed_res))
parser, parsed_res = cloudpickle.loads(parsed_pair)
```
![image](https://github.com/user-attachments/assets/17a8afe3-229e-44fd-8145-239c26cb6721)

**Comment by ptmcg** [comment]
Ooooo, that _does_ look like infinite recursion is happening.

I wonder if this could be addressed by adding `__getstate__` and `__setstate__` methods, as we do to support (de)serialization with `pickle`?

**Comment by ptmcg** [comment]
I guessed right, adding these to `TestClass` makes `cloudpickle` happier:

```python
    def __getstate__(self):
        return (self._parser,)

    def __setstate__(self, state):
        self._parser = state[0]
```

We'll probably want to flesh this out a bit for some of the other attributes, but this is a decent lead.

**Comment by ptmcg** [comment]
I have a pretty strong lead now, and am able to pickle a basic ArithmeticParser and parsed results. I need to get back to work just now, but I'll chase this down this evening. Didn't need to make any changes to pyparsing, all just in plusminus.

**Comment by ptmcg** [comment]
I just committed this code to the repo, if you like you can just copy the plusminus.py file down and see how it works for you. I also added automated version testing with tox, and checked it with Python 3.12 and 3.13 - all checked out.

**Comment by ptmcg** [comment]
Fixed in 0.8.0 release
